### PR TITLE
Add stats --json output

### PR DESCRIPTION
**Pull request by oddsun**
*Closes: #187*

**Describe your changes**
Added json flag and implementation to allow export stats to json file using `memray stats --json ...`. 

**Testing performed**
Added 2 unit tests 
- `tests/unit/test_stats_reporter.py::test_stats_output`: test printing stats to terminal
- `tests/unit/test_stats_reporter.py::test_stats_output_json`: test output json stats

**Additional context**
pycon2023

### CODE REVIEW

**Comment by oddsun** [comment]
new forced-push:
- rebased on latest main
- added description in `docs/stats.rst` 
- added news file for this pr: `news/377.feature.rst`.

**Comment by oddsun** [comment]
new forced-push:
- rebased to head of `main`
- fixed news issue
- fixed mypy errors

**Comment by oddsun** [comment]
it looks like it just needs a newline at end of `news/377.feature.rst`, should I make the change by running `pre-commit run --all-files` and squash commits into 1 again after rebasing on top of latest `main` branch?

File: tests/unit/test_stats_reporter.py
Comment (godlygeek) [inline tests/unit/test_stats_reporter.py:378]: Once `fake_stats` is fixed to pass strings, this will need to be:

```suggestion
        "\t MALLOC: 1013\n"
        "\t REALLOC: 797\n"
        "\t CALLOC: 152\n"
        "\t MMAP: 4\n"
```

File: tests/unit/test_stats_reporter.py
Comment (godlygeek) [inline tests/unit/test_stats_reporter.py:393]: Is this `if` really needed? I think the 3.7 branch works for all versions, and we could just unconditionally do:
```suggestion
    printed = "\n".join(" ".join(x[0]) for x in mocked_print.call_args_list)
```

File: tests/unit/test_stats_reporter.py
Comment (godlygeek) [inline tests/unit/test_stats_reporter.py:447]: Same as above, I think we could unconditionally do:
```suggestion
    actual = json_dump.call_args[0][0]
```

File: src/memray/reporters/stats.py
Comment (godlygeek) [inline src/memray/reporters/stats.py:199]: This should be:
```suggestion
    def _get_allocator_type_distribution(self) -> Iterator[Tuple[str, int]]:
```

File: src/memray/reporters/stats.py
Comment (godlygeek) [inline src/memray/reporters/stats.py:15]: This can be removed
```suggestion
```

**Review by godlygeek — CHANGES_REQUESTED** [review]
I've reviewed this, and spotted a few small things that will need to be fixed.

**Comment by godlygeek** [comment]
On top of those, can you also update your commit message to use your real name in the signed-off-by? Per https://github.com/bloomberg/memray/blob/main/CONTRIBUTING.md#contribution-licensing we need a real name for the DCO signoff, not a pseudonym.

File: tests/unit/test_stats_reporter.py
Comment (oddsun) [reply tests/unit/test_stats_reporter.py:111]: should I fix the type sub as well?

**Comment by oddsun** [comment]
Got it. Do you want me to make the changes and squash into a single commit again and then update the signoff?

**Comment by godlygeek** [comment]
> Got it. Do you want me to make the changes and squash into a single commit again and then update the signoff?

Yes, that would be perfect!

File: tests/unit/test_stats_reporter.py
Comment (godlygeek) [reply tests/unit/test_stats_reporter.py]: It looks like you already did, and that's perfect. Thank you!

**Review by godlygeek — APPROVED** [review]
I've pushed a few more commits here to polish this a little bit: adding some documentation, adding support for `--output` and `--force` like our other subcommands that output files accept, and tweaking the format of the JSON a little bit to make it easier to use.

@pablogsal Since I've edited this pretty heavily now, would you mind giving this a review?

**Comment by pablogsal** [comment]
> I've pushed a few more commits here to polish this a little bit: adding some documentation, adding support for `--output` and `--force` like our other subcommands that output files accept, and tweaking the format of the JSON a little bit to make it easier to use.
> 
> @pablogsal Since I've edited this pretty heavily now, would you mind giving this a review?

Will review this today or monday 👍

**Comment by godlygeek** [comment]
The 3.8 tests are failing because of https://github.com/ipython/ipython/issues/14053 - either the IPython maintainers need to yank the 8.13 release, or we need to land #380 and rebase this.

**Comment by gusmonod** [comment]
> The 3.8 tests are failing because of [ipython/ipython#14053](https://github.com/ipython/ipython/issues/14053) - either the IPython maintainers need to yank the 8.13 release, or we need to land #380 and rebase this.

I've landed #380, we can always remove that workaround if and when the IPython maintainers yank the 8.13 release.

**Review by pablogsal — APPROVED** [review]

**Pull request merged**
